### PR TITLE
Merge attributes into objects

### PR DESCRIPTION
**Pull request by mokaddem**
## TO DO
- Review the code of the valid template collection ``validObjectsFromAttributeTypes`` (Could be improved?)

## Allows to quickly merge multiple Attributes into an Object
- Pick a matching template
- Review before merge
- Perform the merge

mergeAttributes

Solve #4388

## Merging mechanism
- Duplicate the merged attributes
  - originals [O] get deleted flag set or gets hard-deleted if the event has never been published (end of procedure)
  - created [C] have their ``object_relation`` set
- Duplicate the tags attached to the old attribute [O] and attach them to the new ones [C]
- Duplicate sightings of [O] and attach to [C]
- Add a warning regarding references (As the attribute is converted into an object, the meaning of the references might change)

### CODE REVIEW

File: app/Controller/AttributesController.php [hunk 1]
```diff
@@ -1497,11 +1426,11 @@ public function deleteSelected($id = false, $hard = false)
         $successes = array();
         foreach ($attributes as $a) {
             if ($hard) {
-                if ($this->__delete($a['Attribute']['id'], true)) {
+                if ($this->Attribute->__delete($a['Attribute']['id'], true)) {
```
Comment (iglocska) [inline app/Controller/AttributesController.php]: The parameters for __delete changed with the refactor, requires user to be passed along.

File: app/Model/Attribute.php [hunk 2]
```diff
@@ -3909,6 +3908,74 @@ public function editAttribute($attribute, $eventId, $user, $objectId, $log = fal
         return true;
     }
 
+    public function __delete($id, $user, $hard = false)
```
Comment (iglocska) [inline app/Model/Attribute.php]: public function, should be called "delete()"
private would be __delete()
protected would be _delete()

File: app/Model/Attribute.php [hunk 3]
```diff
@@ -3909,6 +3908,74 @@ public function editAttribute($attribute, $eventId, $user, $objectId, $log = fal
...
+    {
+        $this->id = $id;
+        if (!$this->exists()) {
+            return false;
+        }
+        $result = $this->find('first', array(
+            'conditions' => array('Attribute.id' => $id),
+            'fields' => array('Attribute.*'),
```
Comment (iglocska) [inline app/Model/Attribute.php]: 'recursive' => -1,
'contain' => array('Event')

would achieve the same and be cleaner.

File: app/Model/Attribute.php (hunk 2 above)
Comment (iglocska) [reply app/Model/Attribute.php]: However, delete on the model level is already defined, so consider a different name.

File: app/Model/MispObject.php [hunk 4]
```diff
@@ -725,4 +725,147 @@ public function removeOrphanedObjects()
...
+                'ObjectTemplate.name',
+                'COUNT(ObjectTemplateElement.type) as type_count'
+            ),
+            'conditions' => array(
+                'ObjectTemplate.active' => true,
+                'ObjectTemplateElement.type' => $attribute_types
+            ),
+            'joins' => array(
+                array(
+                    'table' => $db->fullTableName($this->ObjectTemplate->ObjectTemplateElement),
```
Comment (iglocska) [inline app/Model/MispObject.php]: Can just replace this with

'table' => 'object_template_elements'

File: app/Model/MispObject.php [hunk 5]
```diff
@@ -725,4 +725,147 @@ public function removeOrphanedObjects()
...
+        if (empty($attributes)) {
+            return array('templates' => array(), 'types' => array());
+        }
+        $attribute_types = array();
+        foreach ($attributes as $i => $attribute) {
+            $attribute_types[$attribute['Attribute']['type']] = 1;
+            $attributes[$i]['Attribute']['object_relation'] = $attribute['Attribute']['type'];
+        }
+        $attribute_types = array_keys($attribute_types);
+        $db = $this->getDataSource();
```
Comment (iglocska) [inline app/Model/MispObject.php]: No need for this, see later.

File: app/Controller/AttributesController.php (hunk 1 above)
Comment (mokaddem) [reply app/Controller/AttributesController.php]: Nice catch!

File: app/Model/MispObject.php (hunk 4 above)
Comment (mokaddem) [reply app/Model/MispObject.php]: Though it was good practice.. ;)

File: app/Model/Attribute.php (hunk 2 above)
Comment (mokaddem) [reply app/Model/Attribute.php]: Called it ``deleteAttribute``

**Comment by iglocska** [comment]
Thanks a lot!

**Pull request merged**